File: Beecrowd/6-Paradigmas/2191-Melhor_Ordem-7/2919.cs
using System;
using System.Collections.Generic;

class Program
{
    static int LowerBound(List<int> list, int target)
    {
        int l = 0, r = list.Count;
        while (l < r)
        {
            int mid = (l + r) / 2;
            if (list[mid] < target) l = mid + 1;
            else r = mid;
        }
        return l;
    }

    static int LIS(List<int> nums)
    {
        List<int> lis = new List<int>();
        foreach (int x in nums)
        {
            int idx = LowerBound(lis, x);
            if (idx == lis.Count) lis.Add(x);
            else lis[idx] = x;
        }
        return lis.Count;
    }

    static void Main()
    {
        List<string> lines = new List<string>();
        string line;
        while ((line = Console.ReadLine()) != null)
        {
            if (line.Trim() != "") lines.Add(line);
        }

        int i = 0;
        while (i + 1 < lines.Count)
        {
            int n = int.Parse(lines[i]);
            i++;
            List<int> nums = new List<int>();

            while (nums.Count < n && i < lines.Count)
            {
                string[] parts = lines[i].Trim().Split(new[] { ' ', '\t' }, StringSplitOptions.RemoveEmptyEntries);
                foreach (var part in parts)
                    nums.Add(int.Parse(part));
                i++;
            }

            Console.WriteLine(LIS(nums));
        }
    }
}

File: Beecrowd/6-Paradigmas/2191-Melhor_Ordem-7/2919.js
const fs = require('fs');
const input = fs.readFileSync(0, 'utf8').split('\n');

function lowerBound(arr, target) {
    let l = 0, r = arr.length;
    while (l < r) {
        let mid = (l + r) >> 1;
        if (arr[mid] < target) l = mid + 1;
        else r = mid;
    }
    return l;
}

function LIS(arr) {
    const lis = [];
    for (let x of arr) {
        const i = lowerBound(lis, x);
        if (i === lis.length) lis.push(x);
        else lis[i] = x;
    }
    return lis.length;
}

let i = 0;
while (i < input.length) {
    while (i < input.length && input[i].trim() === '') i++; // ignora linhas vazias
    if (i >= input.length) break;

    const n = parseInt(input[i++]);
    let nums = [];

    while (nums.length < n && i < input.length) {
        nums.push(...input[i++].trim().split(/\s+/).map(Number));
    }

    console.log(LIS(nums));
}
